server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java

```diff
@@ -274,6 +291,62 @@ public void onReplicationFailure(
         }
     }
 
+    protected void updateVisibleCheckpoint(long replicationId, IndexShard replicaShard) {
+        ShardRouting primaryShard = clusterService.state().routingTable().shardRoutingTable(replicaShard.shardId()).primaryShard();
+
+        final UpdateVisibleCheckpointRequest request = new UpdateVisibleCheckpointRequest(
+            replicationId,
+            replicaShard.routingEntry().allocationId().getId(),
+            primaryShard.shardId(),
+            getPrimaryNode(primaryShard),
+            replicaShard.getLatestReplicationCheckpoint()
+        );
+
+        final TransportRequestOptions options = TransportRequestOptions.builder()
```

Comment:
I'm wondering if we should move all this client code to PrimaryShardReplicationSource and fetch it from SegmentReplicationSourceFactory.  With the implementation of remote store integration the need for the ReplicationSource interface is gone, so we could do this in [refactoring](https://github.com/opensearch-project/OpenSearch/issues/7802) issue.

Reply:
Sure.
 I originally started with the same thought but then realised that i'll have add this logic in 2 places `PrimaryShardReplicationSource` and `RemoteStoreReplicationSource`(or make the `SegmentReplicationSource` abstract). Thus decided against it. 
We can address this with refactor.